### ISSUE TITLE
Elektra should describe use case more clearly

### ISSUE THREAD

**Issue by eschnett** [ElektraInitiative/libelektra#1054]
While looking at libelektra, I was initially confused about what it actually does. Its functionality is described at a higher level ("framework to access configuration parameters"), but this is quite generic.

For example, I write simulation codes that need to access (at run time) parameters describing the simulation to be performed. By just looking at libelektra's summary and Github readme, this seems to be a reasonable application. However, after studying the tutorial and documentation in detail, I find that libelektra is actually not a good match for this use case. The reasons are e.g.:
- libelektra allows dynamically changing the configuration
- libelektra cannot enforce (or check) relations between different configuration key/value pairs
- changing parameters at run time does not make sense for a simulation running in batch mode; in fact, instead configurations need to be immutable and need to be archived for reproducibility

The actual features of libelektra are only described in a single sentence "Elektra serves as a universal and secure framework to access configuration parameters in a global, hierarchical key database." on the Github landing page, before diving into details. The software paper's summary has even less information regarding intended use cases.

I think there is one paragraph missing, describing libelektra at a level that is lower than "framework to access configuration parameters", but higher than e.g. "configuration data can come from any data source". This paragraph could describe that libelektra is dynamic and allows real-time changes, or could give a concrete example for two or three use cases (maybe managing startup scripts on clusters of workstations? managing internet-of-things devices at home?).

**Comment by markus2330** [comment]
Thanks for this very good feedback!

Elektra with some additional plugins should be able to fulfill the needs for your application. You can think of Elektra as a virtual file system for configuration. Because of this, it is quite difficult to come up with an example that cannot be supported: Like you can store any data in a virtual file system (you have only to pick/implement the right concrete file system), you can store any configuration in Elektra (you have only to pick/implement the right plugins). (Of course there are some limitations: e.g. if you cannot transform the configuration to key/value pairs.)

>   libelektra allows dynamically changing the configuration

This can be "disabled" by never calling `kdbGet` in your application again. Or by writing a plugin that rejects every `kdbGet` except the first one for your application.

> libelektra cannot enforce (or check) relations between different configuration key/value pairs

E.g. the [conditionals plugins](https://github.com/ElektraInitiative/libelektra/tree/master/src/plugins/conditionals) can describe relations between key/value pairs.

> instead configurations need to be immutable and need to be archived for reproducibility

This could be ensured via something like the [fcrypt plugin](https://github.com/ElektraInitiative/libelektra/tree/master/src/plugins/fcrypt): by checking the signature of your configuration file you could ensure that the same configuration is passed again.

But you are completely right that we should describe more concrete use cases and examples. Currently `doc/tutorials/application-integration.md` is the only description.

**Comment by markus2330** [comment]
We addressed this issue with following changes:

- We added a longer higher-level description http://tree.libelektra.org/doc/BIGPICTURE.md
- We wrote more about usecases and "who we target" in  http://tree.libelektra.org/doc/WHY.md
- We added more concrete use cases related to research in http://tree.libelektra.org/doc/paper/paper.md

Hopefully you find the paragraph you have been missing there.

But I think Elektra should not hide the fact that there are many things we do not yet know about the abstraction we create. Similar to operating system research, it is a non-trivial issue what the boundaries of the system should be.

To fully answer if Elektra is well suited for your application (immutable, reproducable, non-reconfigurable configuration with enforced relations) we actually need to try it out. We strive to extend the [tutorials](http://tree.libelektra.org/doc/tutorials/) with use cases where Elektra is known to work very nicely. This is an ongoing process because of the immense number of possible use cases.
